fix(federation): remove background jobs when removing trusted servers

kesselb opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

## Summary

Add event listener to remove matching background jobs for trusted server url.

Motivation: When debugging federation locally, I occasionally remove and re-add the server. This can result in multiple background jobs for the same url/target, but with different tokens. I've encountered this issue several times. Therefore, I added an event listener to remove them immediately. I won't backport this change as it has minimal impact on production systems (since those jobs will expire eventually), and it primarily improves the developer experience.


## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

kesselb committed:
fix(federation): remove background jobs when removing trusted servers

Add event listener to remove matching background jobs for trusted server url.

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

nickvergessen commented:
Btw while on it, when another server adds your server as trusted, they run this background job all the time, but as a "recipient" you don't see that someone is asking to get added as a trusted server, neither can you deny it, so the other server gives up, right?

kesselb commented:
> Btw while on it, when another server adds your server as trusted, they run this background job all the time, but as a "recipient" you don't see that someone is asking to get added as a trusted server, neither can you deny it, so the other server gives up, right?

Yes

kesselb committed:
fix: Make DummyJobList.getJobsIterator return an interable instance

iterator_to_array on PHP 8.1 does not accept an array and fails hard with a type error

Signed-off-by: Daniel Kesselberg <mail@danielkesselberg.de>

nickvergessen commented on apps/federation/tests/Listener/TrustedServerRemovedListenerTest.php:
```diff
@@ -0,0 +1,71 @@
+		$job2 = $this->createGetSharedSecretMock();
+		$job3 = $this->createGetSharedSecretMock();
+		$job4 = $this->createGetSharedSecretMock();
+		$this->jobList->add($job1, ['url' => 'https://example.org']);
+		$this->jobList->add($job2, ['url' => 'https://example.net']);
+		$this->jobList->add($job3, ['url' => $url]);
+		$this->jobList->add($job4, ['url' => $url]);
```
should add the other arguments as well, to make it clearer why you don't simply get the job by `['url' => $url]` later on and instead need to iterate?

kesselb replied:
Thank you, that is a good suggestion.
 Update the test and added a brief description to the removeJobsByUrl function.

pabzm commented:
I can't dive into this in the next few days, sorry.

kesselb commented on lib/public/Federation/Events/TrustedServerRemovedEvent.php:
```diff
@@ -14,14 +14,23 @@
+	public function __construct(
+		private readonly string $url,
+		private readonly string $urlHash,
+	) {
```
Done :heavy_check_mark: 
Thanks for the reminder :see_no_evil:
